drivers/regulator/regulator_rpi_pico.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT raspberrypi_core_supply_regulator
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/regulator.h>
+#include <zephyr/dt-bindings/regulator/rpi_pico.h>
+#include <zephyr/sys/linear_range.h>
+#include <zephyr/toolchain.h>
+#include <hardware/regs/vreg_and_chip_reset.h>
+#include <hardware/structs/vreg_and_chip_reset.h>
+
+static const struct linear_range core_ranges[] = {
+	LINEAR_RANGE_INIT(800000u, 0u, 0u, 5u),
+	LINEAR_RANGE_INIT(850000u, 50000u, 6u, 15u),
+};
+
+struct regulator_rpi_pico_config {
+	struct regulator_common_config common;
+	const struct linear_range *ranges;
+	const size_t num_ranges;
+	const uint32_t reg;
+	const bool brown_out_detection;
+	const uint32_t brown_out_threshold;
+};
+
+struct regulator_rpi_pico_data {
+	struct regulator_common_data data;
+};
+
+/*
+ * APIs
+ */
+
+static unsigned int regulator_rpi_pico_count_voltages(const struct device *dev)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+
+	return linear_range_group_values_count(config->ranges, config->num_ranges);
+}
+
+static int regulator_rpi_pico_list_voltage(const struct device *dev, unsigned int idx,
+					   int32_t *volt_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+
+	return linear_range_group_get_value(config->ranges, config->num_ranges, idx, volt_uv);
+}
+
+static int regulator_rpi_pico_set_voltage(const struct device *dev, int32_t min_uv, int32_t max_uv)
+{
+	const struct regulator_rpi_pico_config *config = dev->config;
+	volatile uint32_t *vreg = &(((volatile vreg_and_chip_reset_hw_t *)config->reg)->vreg);
```

Comment:
nit: just make config->reg a `vreg_and_chip_reset_hw_t *` pointer

Reply:
Applied it.